# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd
import seaborn as sns

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(10)

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# Сгруппируем клиентов по количеству детей и посчитаем количество заемщиков в каждой группе
gr_by_childrens = data.groupby('children')['debt'].count()
gr_by_childrens

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

In [23]:
# Найдем количество должников в каждой группе
gr_by_childrens_debt = data.groupby('children')['debt'].sum()
gr_by_childrens_debt

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

In [24]:
# Найдем процентное соотношение должников к общему количеству заемщиков в каждой из групп
debitor_ratio_by_children = (gr_by_childrens_debt / gr_by_childrens * 100).round(1)
debitor_ratio_by_children

children
0    7.5
1    9.2
2    9.5
3    8.2
4    9.8
5    0.0
Name: debt, dtype: float64

In [39]:
# Выведем сводную таблицу с группировкой по количеству детей у заёмщика
children_debt = data.pivot_table(index='children', values='debt', aggfunc=['count', 'sum', 'mean'])
children_debt

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

**Вывод:** 

- Наблюдается несбалансированность групп, доля заемщиков с 3 и более детьми существенно ниже чем, доля заемщиеков без детей или с 1 или 2 детьми.
- Заемщики без детей реже становятся должниками, это происходит в **7.5%** случаев.
- При наличие детей, риск невыплаты кредита **повышается на 0.7 - 2.4 п.п., т.е 8.2 - 9.8%**
- Доля заемщиков с 5 детими очень маленькая, всего 9 заемщиков из 21525, но можно предположить, что риск возникновения задолжности возрастет, т.к, в общем, у заемщиков с детьми больше невозвратов чем у заемщиков без детей

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [27]:
# Сгруппируем заемщиков по семейному положению
gr_by_family_stat = data.groupby('family_status')['family_status'].count()
gr_by_family_stat

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: family_status, dtype: int64

In [28]:
# Найдем количество должников в каждой группе
gr_by_family_stat_debt = data.groupby('family_status')['debt'].sum()
gr_by_family_stat_debt

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

In [29]:
# Найдем процентное соотношение должников к общему кол-ву заемщиуов в каждой группе
debitor_by_family_stat_ratio = (gr_by_family_stat_debt / gr_by_family_stat * 100).round(1)
debitor_by_family_stat_ratio

family_status
Не женат / не замужем    9.8
в разводе                7.1
вдовец / вдова           6.6
гражданский брак         9.3
женат / замужем          7.6
dtype: float64

In [30]:
# Выведем сводную таблицу сгруппированную по семейному положению.
family_stat_debt = data.pivot_table(index='family_status', values='debt', aggfunc=['count', 'sum'])
family_stat_debt['ratio'] = debitor_by_family_stat_ratio
family_stat_debt

count  sum ratio
                        debt debt      
family_status                          
Не женат / не замужем   2796  273   9.8
в разводе               1189   84   7.1
вдовец / вдова           951   63   6.6
гражданский брак        4134  385   9.3
женат / замужем        12261  927   7.6

**Вывод:**
 - Наибольшая доля невозвратов приходится на заемщиков из группы **Не женат/не замужем - 9.8%**
 - Так же высока доля невозвратов в группе **гражданский брак - 9.3%** 
 - Заемщики **состоящие или состоявшие в зарегистрированом браке** реже становятся должниками и процент невозрата займа соствляет **6.6 - 7.6%**

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [31]:
# Сгруппируем заемщиков по категориям из столбца `total_income_category` и посчитаем 
# количество заемщиков в каждой группе
gr_by_income_category = data.groupby('total_income_category')['total_income_category'].count()
gr_by_income_category

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: total_income_category, dtype: int64

In [32]:
# Посчитаем количество должников в каждой из групп
gr_by_income_category_debt = data.groupby('total_income_category')['debt'].sum()
gr_by_income_category_debt 

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

In [33]:
# Найдем процентное соотношение количество должников к общему количеству заемщиков в каждой группе
debitor_by_income_categoty_ratio = (gr_by_income_category_debt / gr_by_income_category * 100).round(1) 
debitor_by_income_categoty_ratio

total_income_category
A    8.0
B    7.1
C    8.5
D    6.0
E    9.1
dtype: float64

In [34]:
# Выведем сводную таблицу сгруппированную по категориям заёмщиков
income_debt = data.pivot_table(index='total_income_category', values='debt', aggfunc=['count', 'sum'])
income_debt['ratio'] = debitor_by_income_categoty_ratio 

# Добавим описание уровня доходов для каждой категории заёмщиков
income_debt['income'] = ['больше 1000000 руб.', '200001 - 1000000 руб.', '50001 - 200000 руб.', '30001 - 50000 руб.', 'меньшне 30000 руб.']
income_debt

count   sum ratio                 income
                        debt  debt                             
total_income_category                                          
A                         25     2   8.0    больше 1000000 руб.
B                       5014   354   7.1  200001 - 1000000 руб.
C                      15921  1353   8.5    50001 - 200000 руб.
D                        349    21   6.0     30001 - 50000 руб.
E                         22     2   9.1     меньшне 30000 руб.

**Вывод:** 
- Наблюдается не сбаланисорованность выборок
- В группах **А** и **E** очень маленькая выборка, **25** и **22** заемщика соответственно, в группе **D** всего **349** заемщиков, что составляет 1.6% от общего числа заёмщиков.
- По группам **B** и **C** можно сделать вывод, что с **ростом дохода** количество невозвратов кредита **уменьшается**

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [35]:
# Сгруппируем заемщиков по целям взятия кредита и посчитаем количество заемщиков в каждой группе
gr_by_purpose_category = data.groupby('purpose_category')['purpose_category'].count()
gr_by_purpose_category

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: purpose_category, dtype: int64

In [36]:
# Посчитаем количество невозвратов в каждой группе
gr_by_purpose_category_debt = data.groupby('purpose_category')['debt'].sum()
gr_by_purpose_category_debt

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

In [37]:
# Посчитаем процентное отношение количества задолжностей к общему числу займов в каждой группе
debitor_by_purpose_category_ratio = (gr_by_purpose_category_debt / gr_by_purpose_category * 100).round(1)
debitor_by_purpose_category_ratio

purpose_category
операции с автомобилем      9.3
операции с недвижимостью    7.3
получение образования       9.3
проведение свадьбы          7.9
dtype: float64

In [38]:
# Выведем сводную таблицу сгруппированную по целям кредита
purpose_debt = data.pivot_table(index='purpose_category', values='debt', aggfunc=['count', 'sum'])
purpose_debt['ratio'] = debitor_by_purpose_category_ratio
purpose_debt

count  sum ratio
                           debt debt      
purpose_category                          
операции с автомобилем     4279  400   9.3
операции с недвижимостью  10751  780   7.3
получение образования      3988  369   9.3
проведение свадьбы         2313  183   7.9

**Вывод:** 
- Наименьшее число невозратов к общему числу займов в группе **операции с недвижимостью - 7.3%** и группе **проведение свадьбы - 7.9%**
- При кредите на **автомобиль** или на **получения образования** вероятность невозрата кредита **возрастает**, доля невозвратов к общему числу займов в группе составляет **9.3%**

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

При выводе информации по пропускам, мы видим, что в столбцах `days_employed` и `total_income` одинаковое количество пропусков. Полагаю, это связано с тем, что есть заёмщики, которые подавали заявки, например онлайн, где не обязательно было заполнять доходы и стаж работы.

Возможно, часть из заемщиков не работают и в связи с этим у них нет трудового стажа и доходов.

Возможно, пропуски связаны с ошибкой при выгрузки данных

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 
это связано с тем, что при заполнении пропусков медианным значением, общее медианное значение не изменится, и медиана более устойчива к выбросам, т.е это не должно сильно отразиться на данных и следовательно на получаемых выводах

### Шаг 4: общий вывод.

Были проанализированы табличные данные по выданым ранее кредитам. Целью анализа данных было выявить влияет ли количество детей, семейное положение, уровень дохода и цели кредита на увеличение риска невыполения обязательст по кредиту.

Перед выполнение анализа данных была проведена следующая работа:
- заполнены пропуски значений в столцах `days_employed` и `total_income`
- удалены аномальные значения в столбце `children`
- удалены дубликаты
- были выделены категории заемщиков по доходу
- быди выделены категории по цели кредита

В ходе проведенного анализа было выявлено:
- наличие детей увеличивает риск невозврата займа
- риск невозрата займа уменьшается, если заёмщик состоит или состоял в зарегистрированном браке
- риск невозрата займа уменьшается, при высоком доходе у заемщика
- при целях кредита *операции с автомобилем* и *получение образования* риск невозврата займа увеличивается

Идеальным клиентом для банка будет является клиет с высоким доходом, без детей, состоящий или состоявшим в зарегистированном браке, который берет кредит для покупки жилья